osu.Game/Beatmaps/WorkingBeatmapCache.cs

```diff
@@ -152,14 +153,28 @@ protected override IBeatmap GetBeatmap()
                         return null;
                     }
 
-                    if (stream.ComputeMD5Hash() != BeatmapInfo.MD5Hash)
+                    string streamMD5 = stream.ComputeMD5Hash();
+                    string streamSHA2 = stream.ComputeSHA2Hash();
+
+                    if (streamMD5 != BeatmapInfo.MD5Hash)
                     {
                         Logger.Log($"Beatmap failed to load (file {BeatmapInfo.Path} does not have the expected hash).", level: LogLevel.Error);
                         return null;
                     }
 
                     using (var reader = new LineBufferedReader(stream))
-                        return Decoder.GetDecoder<Beatmap>(reader).Decode(reader);
+                    {
+                        var beatmap = Decoder.GetDecoder<Beatmap>(reader).Decode(reader);
+
+                        beatmap.BeatmapInfo.MD5Hash = streamMD5;
+                        beatmap.BeatmapInfo.Hash = streamSHA2;
+                        beatmap.BeatmapInfo.Length = beatmap.CalculatePlayableLength();
+                        beatmap.BeatmapInfo.BPM = 60000 / beatmap.GetMostCommonBeatLength();
+                        beatmap.BeatmapInfo.EndTimeObjectCount = beatmap.HitObjects.Count(h => h is IHasDuration);
+                        beatmap.BeatmapInfo.TotalObjectCount = beatmap.HitObjects.Count;
```

Comment:
Copied verbatim from

https://github.com/ppy/osu/blob/775cdc087eda5c1525d763c6fa3d422db0e93f66/osu.Game/Beatmaps/BeatmapUpdater.cs#L66-L69

Maybe could be split to helper. I dunno.